cirq-rigetti/cirq_rigetti/aspen_device.py

```diff
@@ -0,0 +1,514 @@
+
+    def __init__(self, isa: Union[InstructionSetArchitecture, Dict[str, Any]]) -> None:
+        """Initializes a RigettiQCSAspenDevice with its Rigetti QCS `InstructionSetArchitecture`.
+
+        Args:
+            isa: The `InstructionSetArchitecture` retrieved from the QCS api.
+
+        Raises:
+            UnsupportedRigettiQCSQuantumProcessor: If the isa does not define
+            an Aspen device.
+        """
+        if isinstance(isa, InstructionSetArchitecture):
+            self.isa = isa
+        else:
+            self.isa = InstructionSetArchitecture.from_dict(isa)
+
+        if self.isa.architecture.family.lower() != 'aspen':
+            raise UnsupportedRigettiQCSQuantumProcessor(
+                'this integration currently only supports Aspen devices, '
+                f'but client provided a {self.isa.architecture.family} device',
+            )
+        self.quantum_processor = QCSQuantumProcessor(
+            quantum_processor_id=self.isa.name, isa=self.isa
+        )
+
+    def qubits(self) -> List['AspenQubit']:
+        """Return list of `AspenQubit`s within device topology.
+
+        Returns:
+            List of `AspenQubit`s within device topology.
+        """
+        qubits = []
+        for node in self.isa.architecture.nodes:
+            qubits.append(AspenQubit.from_aspen_index(node.node_id))
+        return qubits
+
+    @property
+    def qubit_topology(self) -> nx.Graph:
+        """Return qubit topology indices with nx.Graph.
+
+        Returns:
+            Qubit topology as nx.Graph with each node specified with AspenQubit index.
+        """
+        return self.quantum_processor.qubit_topology()
+
+    @property
+    def _number_octagons(self) -> int:
+        return int(np.ceil(self._maximum_qubit_number / 10))
+
+    @property
+    def _maximum_qubit_number(self) -> int:
+        return max([node.node_id for node in self.isa.architecture.nodes])
+
+    @functools.lru_cache(maxsize=2)
+    def _line_qubit_mapping(self) -> List[int]:
+        mapping: List[int] = []
+        for i in range(self._number_octagons):
+            base = i * 10
+            mapping = mapping + [base + index for index in _forward_line_qubit_mapping]
+        for i in range(self._number_octagons):
+            base = (self._number_octagons - i - 1) * 10
+            mapping = mapping + [base + index for index in _reverse_line_qubit_mapping]
+        return mapping
+
+    def _aspen_qubit_index(self, valid_qubit: cirq.Qid) -> int:
+        if isinstance(valid_qubit, cirq.GridQubit):
+            return _grid_qubit_mapping[valid_qubit]
+
+        if isinstance(valid_qubit, cirq.LineQubit):
+            return self._line_qubit_mapping()[valid_qubit.x]
+
+        if isinstance(valid_qubit, cirq.NamedQubit):
+            return int(valid_qubit.name)
+
+        if isinstance(valid_qubit, (OctagonalQubit, AspenQubit)):
+            return valid_qubit.index
+
+        else:  # pragma: no cover
```

Comment:
please use `# coverage: ignore` - I'm not even sure that it is needed in fact as I'm not sure we process `pragma: no cover`